### ISSUE TITLE
[AI][BUG] PPU builds compile the unavailable device tracer, so the flagos profiler has no device timeline

### ISSUE THREAD

**Issue by lvyufeng** [flagos-ai/Torch-FL#411]
## Issue Type
- [x] Bug Report
- [ ] Feature Request
- [ ] Operator Implementation
- [ ] Platform Support
- [ ] Performance Issue
- [ ] Documentation

## AI Agent Information
- **Agent**: Claude Code CLI
- **Model**: Claude Opus 5
- **Session Context**: Surveying PPU-related work surfaced `docs/reference/platform-capability-matrix.md:59`, which records that "PPU withholds the profiler (its tracer is the unavailable implementation)". That phrasing states a vendor limitation. Tracing the claim through `csrc/CMakeLists.txt` showed the opposite: the PPU build never *selects* the CUPTI tracer, so the "unavailable" tracer is a build-configuration outcome, not something PPU forces.

## Summary
Profiling on PPU silently has no device timeline, and the cause is a build-configuration defect rather than a vendor limitation. `csrc/CMakeLists.txt` selects exactly one vendor `DeviceTracer` per build. Its first branch, `if(FLAGOS_ACCELERATOR STREQUAL "cuda" OR FLAGOS_ACCELERATOR STREQUAL "metax")` (line 38), does not accept `ppu`, and PPU matches no `elseif`, so a PPU build falls into the `else()` at line 68. That branch excludes `cupti_device_tracer.cc` but not `unavailable_device_tracer.cc`, so PPU compiles the stub, whose `available()` returns `false` unconditionally (line 40). The kineto adaptor gates registration on `available()` (`flagos_kineto_profiler.cc:224`), so the PPU profiler registers nothing at all and every trace is CPU-only.

Three artifacts in the tree contradict the code and show this is unintended drift rather than a decision:

1. The comment immediately above the branch (lines 33-34) states "CUDA and **PPU** use the CUPTI-compatible tracer" while the condition on line 38 covers only `cuda` and `metax`.
2. `unavailable_device_tracer.cc:15` documents its own audience as "(Ascend, GCU, MUSA, BPU, TsingMicro)" — PPU is not listed.
3. `tests/integration/profiler_support.py` expects PPU to have device, kernel, runtime, flow, linkage and metadata records (`device = platform != "ascend"`, line 99; `ppu` explicitly listed among the memcpy-capable platforms, line 114). The test tree's expectation and the build output are in direct opposition.

PPU profiling has worked before: `a81d69a` (#119, "fix: restore PPU profiler device activities", 2026-08-17) added PPU-specific handling that is still in the tree — HGPTI result-value handling in `cupti_shim.h:43`, and PPU compatibility-clock and KERNEL-vs-CONCURRENT_KERNEL notes in `cupti_device_tracer.cc:662`, `:688`, `:691`, `:722`. None of it is reachable, because the file no longer compiles into a PPU build.

## Environment (for bug reports)
<details>
<summary>Click to expand environment details</summary>

- **Platform**: PPU (T-Head PPU 810e, CUDA-ABI boxing backend)
- **Python**: 3.12
- **PyTorch**: 2.10.0+cpu relinked to the PPU CUDA-13 libtorch (`torch.version.cuda == "13.0"`)
- **torch_fl**: `main` @ `877afb1`
- **Hardware**: 16 x `PPU-ZW810E`; `PPU_SDK=/usr/local/PPU_SDK` (its `CUDA_SDK` is reachable as both `/usr/local/cuda` and `/usr/local/cuda-13.0`, and carries `extras/CUPTI/include/cupti_activity.h`)

**Build config:**
```bash
bash .github/scripts/set_env_ppu.sh   # exports FLAGOS_ACCELERATOR=ppu, CUDA_HOME=$PPU_SDK/CUDA_SDK
```

**Runtime config:**
```bash
export FLAGOS_ACCELERATOR=ppu
export PPU_SDK=/usr/local/PPU_SDK
```
</details>

## Reproduction
The defect is decided at configure time, so it can be shown without GPUs by replaying the branch over the source list:

```python
import re
from pathlib import Path

CMAKE = Path("csrc/CMakeLists.txt").read_text()
ACCELERATOR = "ppu"  # .github/scripts/set_env_ppu.sh:578

block = CMAKE[CMAKE.index("# Select exactly one vendor-specific"):]
block = block[: block.index("\nendif()")]

# Resolve the if/elseif/else chain the way CMake does.
for arm in re.split(r"\n(?=if\(|elseif\(|else\(\))", block):
    head = arm.splitlines()[0]
    if head.startswith("else()") or (
        head.startswith(("if(", "elseif(")) and f'"{ACCELERATOR}"' in head
    ):
        excluded = set(re.findall(r"profiler/(\w+)", arm))
        tracers = {
            "cupti_device_tracer", "roctracer_device_tracer", "cann_device_tracer",
            "musa_mupti_device_tracer", "gcu_topspti_device_tracer",
            "unavailable_device_tracer",
        }
        print(f"PPU arm: {head!r}")
        print(f"  compiled tracer: {sorted(tracers - excluded)}")
        break
```

Current output on `main` @ `877afb1`:

```
PPU arm: 'else()'
  compiled tracer: ['unavailable_device_tracer']
```

The same replay over all nine accelerators in `cmake/flagos_platforms.json`:

```
cuda         'if(... "cuda" OR ... "metax")'   -> ['cupti_device_tracer']
ppu          'else()'                          -> ['unavailable_device_tracer']   <-- defect
metax        'if(... "cuda" OR ... "metax")'   -> ['cupti_device_tracer']
dcu          'elseif(... "dcu")'               -> ['roctracer_device_tracer']
ascend       'elseif(... "ascend")'            -> ['cann_device_tracer']
gcu          'elseif(... "gcu")'               -> ['gcu_topspti_device_tracer']
musa         'elseif(... "musa")'              -> ['musa_mupti_device_tracer']
tsingmicro   'else()'                          -> ['unavailable_device_tracer']
bpu          'else()'                          -> ['unavailable_device_tracer']
```

The observable consequence, once a PPU wheel is built from this tree:

```python
import torch, torch_fl  # noqa: F401

with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CUDA]) as prof:
    x = torch.randn(1024, 1024, device="flagos:0")
    y = x @ x
    torch.flagos.synchronize()

names = {e.name for e in prof.events()}
print(sorted(names))  # no 'kernel', no 'cuda_runtime', no 'gpu_memcpy'
```

## Expected vs Actual Behavior
**Expected:** A PPU build compiles `cupti_device_tracer.cc`, exactly as the branch comment on `csrc/CMakeLists.txt:33-34` claims, and `tests/integration/test_profiler_parity.py -m main_ops` sees `cuda_runtime` / `gpu_memcpy` / `gpu_memset` / `kernel` records the way `tests/integration/profiler_support.py` requires for PPU.

**Actual:** The PPU build compiles `unavailable_device_tracer.cc`, so `MakeDeviceTracer()` returns a tracer whose `available()` is `false`, the kineto adaptor skips registration entirely, and the trace carries only `['Trace', 'cpu_op']`.

## Root Cause Analysis
`06bf923` (#134, "feat: add complete MetaX profiler tracing", 2026-08-21) rewrote the tracer-selection chain. Before it, the chain was two arms:

```cmake
if(ACCELERATOR STREQUAL "dcu")
  list(FILTER SOURCE_FILES EXCLUDE REGEX ".*/profiler/cupti_device_tracer\\.cc$")
  ...
else()
  list(FILTER SOURCE_FILES EXCLUDE REGEX ".*/profiler/roctracer_device_tracer\\.cc$")
  ...
endif()
```

PPU matched no arm and reached `else()`, which excluded only the non-CUDA tracers — so PPU compiled the CUPTI tracer. `#134` replaced this with a five-arm `FLAGOS_ACCELERATOR` ladder whose first arm enumerates `cuda` and `metax` only, *and* added `cupti_device_tracer.cc` to the `else()` exclusions. Both edits together moved PPU from the CUPTI tracer to the unavailable stub, in a commit whose stated scope was MetaX. The misleading "CUDA and PPU use the CUPTI-compatible tracer" comment was written in that same commit: it records the intent, while the condition below it implements only `cuda` + `metax`.

A byte-level replay of the branch over every commit that touched `csrc/CMakeLists.txt` confirms `06bf923` is the only transition, and that it predates `9d84cc9` (#341, the `ACCELERATOR` → `FLAGOS_ACCELERATOR` rename) — #341 renamed the variable in an already-broken chain and did not cause the break.

The PPU-specific CUPTI code written by `a81d69a` (#119) four days earlier shows the tracer was working on PPU: that commit shipped with "Tested: pytest tests/integration/test_profiler_parity.py -v -m main_ops (7 passed)" on PPU hardware.

`.github/configs/ppu.yml:160-170` has the profiler parity step commented out with the rationale that "PPU CUPTI Activity API is defective: `CuptiActivityApi.cpp:298` logs 'bufferCompleted called with unknown buffer'". That exclusion landed in `b3e7c96` (#91, 2026-08-25) — **four days after `#134` had already removed our tracer from the PPU build**. Two things follow, and the issue is careful to separate them:

- The cited log line is emitted by PPU's bundled libtorch's *own* CUPTI activity implementation (`CuptiActivityApi.cpp`), not by `torch_fl`'s tracer, which `dlopen`s CUPTI directly through `cupti_shim.h` and never touches torch's activity API. So that evidence concerns the torch-level CUPTI path and cannot have been evidence about `torch_fl`'s tracer.
- The recorded symptom — a flagos trace containing only `['Trace', 'cpu_op']` — is exactly what a not-compiled-in tracer produces, so it does not discriminate between the two causes. The exclusion was never re-measured against a PPU build that actually includes the CUPTI tracer.

Nothing catches the regression: the parity step is excluded from PPU CI, and no test asserts the tracer ladder covers every accelerator.

The defect has since been promoted into documentation. `docs/reference/platform-capability-matrix.md:59` (from `#405`, "PPU withholds the profiler (its tracer is the unavailable implementation)") describes the build outcome as a platform property. Per `CLAUDE.md`, support must not be inferred from routing or build configuration alone, and this is that inference written down as fact.

## Proposed Solution
1. **Add `ppu` to the CUPTI arm** of `csrc/CMakeLists.txt:38`, so the condition matches its own comment:
   ```cmake
   if(FLAGOS_ACCELERATOR STREQUAL "cuda"
      OR FLAGOS_ACCELERATOR STREQUAL "metax"
      OR FLAGOS_ACCELERATOR STREQUAL "ppu")
   ```
   The header search already works for PPU: `cupti_activity.h` resolves at `/usr/local/cuda/extras/CUPTI/include`, and `/usr/local/cuda` is a symlink to `$PPU_SDK/CUDA_SDK`.
2. **Remove the dead `FLAGOS_HAVE_CUPTI` definition.** `csrc/CMakeLists.txt:233-235` guards it with the same wrong accelerator set, but a full-tree grep finds no reader — it is defined and never consumed, by either the sources or the Python tree. Deleting it removes a second copy of the same mistaken platform list.
3. **Add a regression test** that asserts the ladder picks the intended tracer for every accelerator in `cmake/flagos_platforms.json`. That is a pure text check with no GPU or build needed, in the style of `tests/unit/test_conf_registration_consistency.py` and the existing `csrc/CMakeLists.txt` parsing in `tests/unit/test_setup_build.py`. It would have failed on `#134` and is the only thing that will notice the next platform added to the table.
4. **Re-measure, then correct the records, from measurement.** Rebuild the PPU wheel with the tracer restored and run `tests/integration/test_profiler_parity.py -m main_ops` on PPU hardware. Then, whichever way it comes out:
   - update `docs/reference/platform-capability-matrix.md:59` to state what was measured, rather than inferring the profiler's capability from the build configuration; and
   - re-add the parity step in `.github/configs/ppu.yml:160-170`, or retain the exclusion with a rationale that reflects a fresh measurement.

   If the restored tracer still produces no device records, the vendor-side hypothesis in that comment becomes live again, but it will then be the *conclusion* of a measurement on a correctly-built wheel rather than an inference from a build that never included the tracer.

## Verification Plan
- [ ] Unit test: new test asserting the `csrc/CMakeLists.txt` tracer ladder resolves to the intended tracer for all nine accelerators in `cmake/flagos_platforms.json`.
- [ ] Unit test: `pytest tests/unit/test_setup_build.py tests/unit/test_conf_registration_consistency.py -v` stays green.
- [ ] Integration test: rebuild the PPU wheel with the fix and run `PYTHONPATH=$(pwd) python -m pytest tests/integration/test_profiler_parity.py -v -m main_ops` on PPU 810e, where it must produce `cuda_runtime` / `gpu_memcpy` / `gpu_memset` / `kernel` records for the flagos device.
- [ ] Manual verification: build one PPU wheel each from before and after the change and confirm the compiled object list contains `cupti_device_tracer.cc.o` only after the change (`build/` link command, or `nm` on the built extension for the tracer's symbols).
- [ ] Manual verification: re-run the closed-loop check on a PPU host that the "PPU" row of `docs/reference/platform-capability-matrix.md` matches the built wheel.

## Context & Investigation
Commands run and artifacts read, in order:

- Enumerated every commit touching `csrc/CMakeLists.txt` and replayed its tracer-selection arms to determine, per commit, whether a PPU build receives `cupti_device_tracer.cc`. Only `06bf923` (#134) flips it, from `True` to `False`; it stays `False` through `9d84cc9` (#341) and to the present `877afb1`.
- Read the CMake block at `fedfd18` (pre-#341) and confirmed the comment "CUDA and PPU use the CUPTI-compatible tracer" is present with a `cuda OR metax` condition beneath it.
- `git show 06bf923 -- csrc/CMakeLists.txt` to confirm `#134` introduced both the five-arm ladder and the `cupti_device_tracer.cc` exclusion in the `else()` arm.
- `git log -S "HGPTI"` and `git log -S "PPU's CUPTI compatibility"` over `csrc/profiler/` — both resolve to `a81d69a` (#119), whose message records a passing PPU parity run.
- `git log -S "CuptiActivityApi.cpp:298"` over `.github/configs/ppu.yml` — resolves to `b3e7c96` (#91), dated after `#134`.
- Read `csrc/profiler/unavailable_device_tracer.cc`, `csrc/profiler/flagos_kineto_profiler.cc:220-232`, `tests/integration/profiler_support.py:88-135`, and `docs/reference/platform-capability-matrix.md`.
- Confirmed `PPU_SDK` headers: `/usr/local/PPU_SDK/CUDA_SDK/extras/CUPTI/include/cupti_activity.h` exists and `/usr/local/cuda-13.0` is a symlink to `/usr/local/PPU_SDK/CUDA_SDK`, so the CUPTI search paths in `csrc/CMakeLists.txt:223-229` already resolve on a PPU host.
- Searched the whole open issue list (40 issues) for PPU: no open issue is specific to PPU, and no PPU label exists, so this defect was untracked.

Searched for duplicates: no open or closed issue describes the tracer not being compiled in. Closed #92 ("PPU profiler device activities missing") was closed as not-a-repo-defect, attributing the symptom to the vendor CUPTI library — an attribution this analysis contradicts, since `a81d69a` (#119) closed the same symptom on the same day by fixing the repository side.

## Related Code Locations
- `csrc/CMakeLists.txt:33-34` — comment claiming PPU uses the CUPTI-compatible tracer.
- `csrc/CMakeLists.txt:38` — the `cuda OR metax` condition that omits `ppu`.
- `csrc/CMakeLists.txt:68-74` — the `else()` arm a PPU build reaches; excludes `cupti_device_tracer.cc`, keeps `unavailable_device_tracer.cc`.
- `csrc/CMakeLists.txt:233-235` — dead `FLAGOS_HAVE_CUPTI` definition guarded by the same incomplete accelerator set.
- `csrc/profiler/unavailable_device_tracer.cc:15` — audience comment listing every platform but PPU.
- `csrc/profiler/unavailable_device_tracer.cc:40` — `available()` returning `false`, which suppresses registration.
- `csrc/profiler/flagos_kineto_profiler.cc:224` — the `tracer_->available()` gate.
- `csrc/profiler/cupti_shim.h:43`, `csrc/profiler/cupti_device_tracer.cc:662,688,691,722` — live PPU-specific CUPTI code that no PPU build compiles.
- `tests/integration/profiler_support.py:99,114` — PPU expected to provide device and memcpy records.
- `.github/configs/ppu.yml:160-170` — the profiler parity step excluded from PPU CI.
- `docs/reference/platform-capability-matrix.md:59` — the PPU profiler limitation recorded as a platform property.
- `cmake/flagos_platforms.json` — authoritative accelerator list the proposed regression test should iterate.

## Checklist - AI Agents MUST Complete All
- [x] I have provided complete environment information
- [x] I have included a minimal, self-contained reproducer
- [x] I have included full error output with traceback
- [x] I have analyzed the root cause (not just symptoms)
- [x] I have proposed a specific solution with implementation approach
- [x] I have identified affected code locations with line numbers
- [x] I have described how to verify the fix
- [x] I have checked for duplicate issues
- [x] All text is in **English** (required per CLAUDE.md)
- [x] Code follows project conventions (checked existing code style)
